readme.md
## Description
This adds a plugin for precise, this is the official mycroft wake word 
engine and is supported out of the box by core, however i have some issues 
with the official implementation that i try to solve with this

- the downloading of binaries is messy, i prefer that the download blocks 
  instead of reloading the listener
- mycroft-core keeps replacing and redownloading binary whenever you test a 
  new model from a different version, i want to support having both versions 
  side by side
  - i support [multiple wakeword at same time](https://github.com/MycroftAI/mycroft-core/pull/1233)
- xdg compliant paths
- allow overriding binary used
- fix recording of wake words setting, this is derived from phonemes and 
  badly implemented in core, so you end up with issues like getting 0.2 second 
  recordings if you don't set the phonemes (that are not used by precise)
- i prefer hardcoded static and explicit urls for binary to be downloaded

The "plugins" are pip install-able modules that provide new engines for mycroft

more info in the [docs](https://mycroft-ai.gitbook.io/docs/mycroft-technologies/mycroft-core/plugins)

## Install

`pip install ovos-ww-plugin-precise`

Configure your wake word in mycroft.conf

```json
 "listener": {
      "wake_word": "android"
 },
 "hotwords": {
    "android": {
        "module": "ovos-ww-plugin-precise",
        "version": "0.3",
        "model": "path/to/your/model"
    }
  }
 
```

Advanced configuration

```json
 "listener": {
      "wake_word": "android"
 },
 "hotwords": {
    "android": {
        "module": "ovos-ww-plugin-precise",
        "trigger_level": 3,
        "sensitivity": 0.5,
        "version": "0.3",
        "model": "path/or/url/to/your/model",
        "binary_path": "path/to/your/binary/eg/using/tflite",
        "expected_duration": 3
    }
  }
 
```


- `binary_path` if you want to use your own binary, eg for use with 
  [tflite branch](https://github.com/MycroftAI/mycroft-precise/pull/141)
- `version` defaults to 0.2, only 0.2 and 0.3 supported (ignored if binary path is set)
- `model` full path to your model file, if it's an url the model will be 
  downloaded to a XDG directory
- `trigger_level` Higher = more delay & less sensitive
- `sensitivity`  Higher = more sensitive
- `expected_duration` defaults to 3 seconds (max value), this is the time 
  used for [saving wake word samples](https://github.com/MycroftAI/mycroft-core/blob/4c84f66e15a361d9f3d650def1ba97fa80506456/mycroft/configuration/mycroft.conf#L160)

CHANGELOG.md
# Changelog

## [V0.1.1a2](https://github.com/OpenVoiceOS/ovos-ww-plugin-precise/tree/V0.1.1a2) (2022-02-28)

[Full Changelog](https://github.com/OpenVoiceOS/ovos-ww-plugin-precise/compare/V0.1.1a1...V0.1.1a2)

## [V0.1.1a1](https://github.com/OpenVoiceOS/ovos-ww-plugin-precise/tree/V0.1.1a1) (2022-02-28)

[Full Changelog](https://github.com/OpenVoiceOS/ovos-ww-plugin-precise/compare/a6d9b7c877d91626d891746a5d787bab42fd03aa...V0.1.1a1)



\* *This Changelog was automatically generated by [github_changelog_generator](https://github.com/github-changelog-generator/github-changelog-generator)*


test/license_tests.py
import unittest
from pprint import pprint

from lichecker import LicenseChecker

# these packages dont define license in setup.py
# manually verified and injected
license_overrides = {
    "kthread": "MIT",
    'yt-dlp': "Unlicense",
    'pyxdg': 'GPL-2.0',
    'ptyprocess': 'ISC',
    'PyAudio': 'MIT',
    'petact': 'MIT',
    "sonopy": "Apache-2.0",
    "precise-runner": "Apache-2.0",
    'psutil': 'BSD3'
}
# explicitly allow these packages that would fail otherwise
whitelist = ["petact", "precise-runner"]

# validation flags
allow_nonfree = False
allow_viral = False
allow_unknown = False
allow_unlicense = True
allow_ambiguous = False

pkg_name = "ovos-ww-plugin-precise"


class TestLicensing(unittest.TestCase):
    @classmethod
    def setUpClass(self):
        licheck = LicenseChecker(pkg_name,
                                 license_overrides=license_overrides,
                                 whitelisted_packages=whitelist,
                                 allow_ambiguous=allow_ambiguous,
                                 allow_unlicense=allow_unlicense,
                                 allow_unknown=allow_unknown,
                                 allow_viral=allow_viral,
                                 allow_nonfree=allow_nonfree)
        print("Package", pkg_name)
        print("Version", licheck.version)
        print("License", licheck.license)
        print("Transient Requirements (dependencies of dependencies)")
        pprint(licheck.transient_dependencies)
        self.licheck = licheck

    def test_license_compliance(self):
        print("Package Versions")
        pprint(self.licheck.versions)

        print("Dependency Licenses")
        pprint(self.licheck.licenses)

        self.licheck.validate()


scripts/bump_major.py
import fileinput
from os.path import join, dirname


version_file = join(dirname(dirname(__file__)), "ovos_ww_plugin_precise", "version.py")
version_var_name = "VERSION_MAJOR"
minor_var_name = "VERSION_MINOR"
build_var_name = "VERSION_BUILD"
alpha_var_name = "VERSION_ALPHA"

with open(version_file, "r", encoding="utf-8") as v:
    for line in v.readlines():
        if line.startswith(version_var_name):
            version = int(line.split("=")[-1])
            new_version = int(version) + 1

for line in fileinput.input(version_file, inplace=True):
    if line.startswith(version_var_name):
        print(f"{version_var_name} = {new_version}")
    elif line.startswith(minor_var_name):
        print(f"{minor_var_name} = 0")
    elif line.startswith(build_var_name):
        print(f"{build_var_name} = 0")
    elif line.startswith(alpha_var_name):
        print(f"{alpha_var_name} = 0")
    else:
        print(line.rstrip('\n'))


scripts/bump_minor.py
import fileinput
from os.path import join, dirname


version_file = join(dirname(dirname(__file__)), "ovos_ww_plugin_precise", "version.py")
version_var_name = "VERSION_MINOR"
build_var_name = "VERSION_BUILD"
alpha_var_name = "VERSION_ALPHA"

with open(version_file, "r", encoding="utf-8") as v:
    for line in v.readlines():
        if line.startswith(version_var_name):
            version = int(line.split("=")[-1])
            new_version = int(version) + 1

for line in fileinput.input(version_file, inplace=True):
    if line.startswith(version_var_name):
        print(f"{version_var_name} = {new_version}")
    elif line.startswith(build_var_name):
        print(f"{build_var_name} = 0")
    elif line.startswith(alpha_var_name):
        print(f"{alpha_var_name} = 0")
    else:
        print(line.rstrip('\n'))


jarbas_wake_word_plugin_precise/__init__.py
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
import distutils.spawn
from ovos_plugin_manager.templates.hotwords import HotWordEngine
from ovos_utils.log import LOG
from os.path import join, isfile, expanduser, isdir
from petact import install_package
import platform
from xdg import BaseDirectory
from precise_runner import PreciseRunner, PreciseEngine, ReadWriteStream


class PreciseHotwordPlugin(HotWordEngine):
    """Precise is the default wake word engine for Mycroft.

    Precise is developed by Mycroft AI and produces quite good wake word
    spotting when trained on a decent dataset.
    """

    def __init__(self, key_phrase="hey mycroft", config=None, lang="en-us"):
        super().__init__(key_phrase, config, lang)
        self.expected_duration = self.config.get("expected_duration") or 3
        self.has_found = False
        self.stream = ReadWriteStream()

        self.trigger_level = self.config.get('trigger_level', 3)
        self.sensitivity = self.config.get('sensitivity', 0.5)
        self.version = str(self.config.get("version", "0.2"))

        dist_exe = distutils.spawn.find_executable("precise-engine")
        precise_exe = self.config.get("binary_path", dist_exe)
        model = self.config.get('model')

        if not precise_exe:
            precise_exe = self.get_binary(self.version)

        if model and model.startswith("http"):
            model = self.download_model(model)

        if not model or not isfile(expanduser(model)):
            raise ValueError("Model not found")

        self.precise_model = expanduser(model)

        self.runner = PreciseRunner(
            PreciseEngine(precise_exe, self.precise_model),
            self.trigger_level, self.sensitivity,
            stream=self.stream, on_activation=self.on_activation,
        )
        self.runner.start()

    @staticmethod
    def get_binary(version="0.2"):
        xdg_folder = BaseDirectory.xdg_data_home
        base_url = "https://github.com/MycroftAI/mycroft-precise/releases/download/"
        if version == "0.2":
            base_url += "v0.2.0/precise-engine_0.2.0_{arch}.tar.gz"
            folder = join(xdg_folder, 'precise02')
        elif version == "0.3":
            base_url += "v0.3.0/precise-engine_0.3.0_{arch}.tar.gz"
            folder = join(xdg_folder, 'precise03')
        else:
            LOG.info("Supported versions are 0.2 and 0.3, "
                     "please provide a path to the precise binary")
            raise ValueError("Unsupported Precise Version")

        if not isdir(folder):
            LOG.info("Downloading binary for precise {v}".format(v=version))
            LOG.info("This might take a while")
            arch = platform.machine()
            install_package(base_url.format(arch=arch), folder)
            LOG.info("Binary downloaded")

        precise_exe = join(folder, 'precise-engine', 'precise-engine')
        return precise_exe

    def download_model(self, url):
        xdg_folder = BaseDirectory.xdg_data_home
        if self.version == "0.2":
            folder = join(xdg_folder, 'precise02models')
        elif self.version == "0.3":
            folder = join(xdg_folder, 'precise03models')
        else:
            LOG.info("Supported versions are 0.2 and 0.3, "
                     "please provide a path to the precise binary")
            raise ValueError("Unsupported Precise Version")

        name = url.split("/")[-1].split(".")[0] + ".pb"
        model_path = join(folder, name)
        if not isfile(model_path):
            LOG.info("Downloading model for precise {v}".format(
                v=self.version))
            LOG.info(url)
            install_package(url, folder)
            LOG.info(f"Model downloaded to {model_path}")

        return model_path

    def on_activation(self):
        self.has_found = True

    def update(self, chunk):
        self.stream.write(chunk)

    def found_wake_word(self, frame_data):
        if self.has_found:
            self.has_found = False
            return True
        return False

    def stop(self):
        if self.runner:
            self.runner.stop()
